components/leds.py

```diff
@@ -0,0 +1,403 @@
+import math
+import wpilib
+import time
+from enum import Enum, auto
+from typing import List, Tuple
+from utilities.scalers import scale_value
+import random
+
+
+MAX_BRIGHTNESS = 100  # Between 0-255 of Value on HSV scale
+
+
+class LedColours(Enum):
+    # Use HSV to get nicer fading, hues are 0-180 so half usual hue
+    RED = (0, 255, MAX_BRIGHTNESS)
+    ORANGE = (20, 255, MAX_BRIGHTNESS)
+    YELLOW = (30, 255, MAX_BRIGHTNESS)
+    PINK = (150, 255, MAX_BRIGHTNESS)
+    BLUE = (120, 255, MAX_BRIGHTNESS)
+    CYAN = (90, 255, MAX_BRIGHTNESS)
+    GREEN = (60, 255, MAX_BRIGHTNESS)
+    WHITE = (0, 0, MAX_BRIGHTNESS)
+    VIOLET = (300, 45, MAX_BRIGHTNESS)
+    OFF = (0, 0, 0)
+
+
+class DisplayType(Enum):
+    PACMAN = auto()
+    RAINBOW = auto()
+    SOLID = auto()
+    PULSE = auto()
+    FLASH = auto()
+    ALTERNATING = auto()
+    HALF_HALF = auto()
+    MORSE = auto()
+
+
+class RobotState(Enum):
+    PICKED_UP_PIECE = auto()
+    LOOKING_FOR_PIECE = auto()
+    OTHER = auto()
+
+
+class Piece(Enum):
+    CONE = auto()
+    CUBE = auto()
+    NONE = auto()
+
+
+class PickupFromSide(Enum):
+    LEFT = auto()
+    RIGHT = auto()
+
+    def colour(self) -> LedColours:
+        if self._value_ == 1:
+            return LedColours.GREEN
+        elif self._value_ == 2:
+            return LedColours.RED
+        return LedColours.PINK
```

Comment:
Consider putting three colours in the enum's value instead, then we won't have this bizarre unreachable code.